daml-lf/data/src/main/scala/com/daml/TreeMap.scala

```diff
@@ -5,30 +5,43 @@ package com.daml.lf
 package data
 
 import scala.collection.immutable
+import scala.collection.mutable.ArrayBuffer
 
 object TreeMap {
 
-  import Ordering.Implicits._
-
   // Linear time builder for immutable TreeMap.
-  // Requires the entries to be strictly ordered, crashes if not.
+  // Requires the entries to be ordered, crashes if not.
+  // Like Normal TreeMap builder, in case of duplicate key, the last wins
   // Relies on the fact that immutable.TreeMap.from is linear when its argument
   // is a SortedMap with the same ordering.
   @throws[IllegalArgumentException]
-  def fromOrderedEntries[K: Ordering, V](entries: Iterable[(K, V)]): immutable.TreeMap[K, V] =
+  def fromOrderedEntries[K, V](entries: Iterable[(K, V)])(implicit order: Ordering[K]): immutable.TreeMap[K, V] =
     if (entries.isEmpty)
       immutable.TreeMap.empty
     else {
 
-      val _ = entries.tail.foldLeft(entries.head._1) { case (previousKey, (k, _)) =>
-        require(previousKey < k, "the entries are not strictly ordered")
-        k
+      val buffer = new ArrayBuffer[(K, V)]()
+      buffer.sizeHint(entries.knownSize)
+      val it = entries.iterator
+      var previous = it.next()
+
+      while(it.hasNext) {
+        val next = it.next()
+        order.compare(previous._1, next._1) match {
+          case i if i < 0 =>
```

Comment:
Suggestion: you can also use `oder.compare(...).signum` to ensure that the negative one is -1, which would make this pattern matching slightly more readable.